File: core/src/com/g3ida/withflyingcolours/screens/GameScreen.java
package com.g3ida.withflyingcolours.screens;

import com.badlogic.ashley.core.PooledEngine;
import com.badlogic.gdx.ScreenAdapter;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.utils.viewport.ExtendViewport;
import com.badlogic.gdx.utils.viewport.Viewport;
import com.g3ida.withflyingcolours.core.AssetsLoader;
import com.g3ida.withflyingcolours.core.SceneMapper;
import com.g3ida.withflyingcolours.core.scripts.Player;

import games.rednblack.editor.renderer.SceneLoader;
import games.rednblack.editor.renderer.resources.ResourceManager;
import games.rednblack.editor.renderer.utils.ItemWrapper;

public class GameScreen extends ScreenAdapter {

    private SceneLoader _sceneLoader;
    private ResourceManager _resourceManager;

    private Viewport _viewport;
    private OrthographicCamera _camera;

    private AssetsLoader _assetsLoader;

    private PooledEngine _engine;

    public GameScreen() {

        _assetsLoader = new AssetsLoader("project.dt");

        _resourceManager = _assetsLoader.load();
        _sceneLoader = new SceneLoader(_resourceManager);

        _engine = _sceneLoader.getEngine();

        _camera = new OrthographicCamera();
        _viewport = new ExtendViewport(13, 7, _camera);

        _sceneLoader.loadScene("MainScene", _viewport);

        SceneMapper sceneMapper = new SceneMapper();
        sceneMapper.mapScripts(_sceneLoader);
    }

    @Override
    public void resize(int width, int height) {
        super.resize(width, height);
        _viewport.update(width, height);

        if (width != 0 && height != 0) {
            _sceneLoader.resize(width, height);
        }
    }

    @Override
    public void dispose() {
        _assetsLoader.dispose();
        super.dispose();
    }

    public void update (float deltaTime) {
        _camera.update();
    }

    @Override
    public void render (float delta) {
        super.render(delta);
        update(delta);

        _viewport.apply();
        _engine.update(delta);
    }

    @Override
    public void pause () {

    }
}

File: core/src/com/g3ida/withflyingcolours/screens/GameScreen.kt
package com.g3ida.withflyingcolours.screens

import com.artemis.World
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.utils.viewport.ExtendViewport
import com.badlogic.gdx.utils.viewport.Viewport
import games.rednblack.editor.renderer.SceneLoader
import com.g3ida.withflyingcolours.core.common.AssetsLoader
import games.rednblack.editor.renderer.SceneConfiguration
import com.g3ida.withflyingcolours.core.common.GameSettings
import com.g3ida.withflyingcolours.core.ecs.systems.*
import com.g3ida.withflyingcolours.core.SceneMapper
import com.g3ida.withflyingcolours.utils.extensions.toSceneLoader
import com.g3ida.withflyingcolours.utils.extensions.withResourceRetriever
import com.g3ida.withflyingcolours.utils.extensions.withSystems
import ktx.app.KtxScreen

class GameScreen : KtxScreen {
    private val mSceneLoader: SceneLoader
    private val mViewport: Viewport
    private val mCamera: OrthographicCamera
    private val mAssetsLoader: AssetsLoader = AssetsLoader("project.dt")
    private val mEngine: World

    override fun resize(width: Int, height: Int) {
        super.resize(width, height)
        mViewport.update(width, height)
        if (width != 0 && height != 0) {
            mSceneLoader.resize(width, height)
        }
    }

    override fun dispose() {
        mAssetsLoader.dispose()
        super.dispose()
    }

    fun update() {
        mCamera.update()
    }

    override fun render(delta: Float) {
        super.render(delta)
        update()
        mViewport.apply()
        mEngine.setDelta(delta)
        mEngine.process()
    }

    override fun pause() {}

    init {
        val resourceManager = mAssetsLoader.load()
        mSceneLoader = SceneConfiguration()
            .withSystems(
                CameraSystem(0f, 20f, 0f, 7f),
                EventListenerSystem(),
                PlayerAnimationSystem(),
                ColorPlatformRenderingSystem()
            )
            .withResourceRetriever(resourceManager)
            .toSceneLoader()

        mEngine = mSceneLoader.engine
        mCamera = OrthographicCamera()
        mViewport = ExtendViewport(13f, 7f, mCamera)
        GameSettings.mainViewPort = mViewport
        mSceneLoader.loadScene("MainScene", mViewport)
        SceneMapper(mSceneLoader)
    }
}